### PR TITLE
Issue2874 Proposed Fix

### LINKED ISSUE #2874: get_preds fails when passed a dataloader where the batchsize does not evenly divide the number of items
Please confirm you have the latest versions of fastai, fastcore, fastscript, and nbdev prior to reporting a bug (delete one): YES

**Describe the bug**
See the forum item: https://forums.fast.ai/t/learn-get-preds-batch-size-and-number-of-items-mismatch-error/79649

If get_preds receives a dataloader where batchsize does not evenly divide the number of items, then it fails with an error:
IndexError: index 98 is out of bounds for dimension 0 with size 96

**To Reproduce**

Call get_preds with a dataloader containing a number of items that is not an integer multiple of the batchsize.

**Expected behavior**
Get_preds should return predictions for all if the items in the dataloader.

**Error with full stack trace**
See the previously reported item on the forum. 
```
```

**Additional context**

### PR DESCRIPTION
**Pull request by kevinbird15**
Closes #2874 

I think the question here is whether we want to create a new dl on Learner.get_preds if a dl is passed.  I am leaning towards it being a good idea since we do create a new dl if a train dataloader is passed and make it work properly

### CODE REVIEW

File: nbs/13a_learner.ipynb [hunk 1]
```diff
@@ -103,10 +103,10 @@
     "    valid_ds = get_data(n_valid)\n",
     "    device = default_device() if cuda else None\n",
     "    if tfmdDL:\n",
-    "        train_dl = TfmdDL(train_ds, bs=bs, shuffle=True, num_workers=0)\n",
+    "        train_dl = TfmdDL(train_ds, bs=bs, shuffle=True, num_workers=0, drop_last=True)\n",
```
Comment (kevinbird15) [inline nbs/13a_learner.ipynb:106]: added `drop_last=True` to the synth_dbunch to more accurately represent a real dls.  Would it be better to make this an argument that defaults to True instead?

**Comment by kevinbird15** [comment]
Thankfully the tests caught an interesting use-case of Learner.get_preds that I hadn't considered.  

Currently the get_preds does the DataLoader check by checking for the existence of a `len`: 
https://github.com/fastai/fastai/blob/681ae88211e0bd7d90259c65ce008fa227d02dfa/fastai/learner.py#L243-L245

Because of this, the LMLearner uses this to pass a list to get_preds rather than a true DataLoader.  
https://github.com/fastai/fastai/blob/681ae88211e0bd7d90259c65ce008fa227d02dfa/fastai/text/learner.py#L158-L161

What I am trying to track down currently is why this works in get_preds?  
I *think* something must happen in self._do_epoch_validate that allows this behavior.
https://github.com/fastai/fastai/blob/681ae88211e0bd7d90259c65ce008fa227d02dfa/fastai/learner.py#L253

**Comment by kevinbird15** [comment]
Still a little more work to get this fully working. I notice that Tabular show_results seems to be giving some issues with the changes but will explore more tomorrow

**Comment by kevinbird15** [comment]
Getting closer on this.  I found the reason that Tabular is failing.  It is because `reorder` in get_preds defaults to True and before since we were passing a list, it didn't have a get_idxs attribute so this wouldn't have executed, but since I am making it a proper DataLoader, it gets into this section of code. 
https://github.com/fastai/fastai/blob/681ae88211e0bd7d90259c65ce008fa227d02dfa/fastai/learner.py#L260

So far, what I've found is that text handles this properly and looks like the input is a tensor 
```
ipdb> (t_)
TensorText([[[2.1220e-02, 9.1441e-07, 7.8817e-03,  ..., 2.0213e-05,
          3.1001e-05, 3.1001e-05],
         [1.8548e-02, 5.8384e-08, 2.9509e-03,  ..., 7.1964e-06,
          4.1149e-06, 4.1149e-06],
         [2.3482e-02, 3.5561e-08, 5.9080e-04,  ..., 2.9862e-06,
          2.2433e-06, 2.2433e-06],
         [6.7106e-04, 1.3014e-08, 8.7557e-05,  ..., 8.7888e-07,
          7.7342e-07, 7.7342e-07],
         [8.5899e-04, 7.3736e-09, 4.2166e-05,  ..., 5.1746e-07,
          5.1974e-07, 5.1974e-07],
         [6.1429e-03, 6.5549e-09, 6.1269e-06,  ..., 1.1527e-06,
          5.2226e-07, 5.2226e-07]]])
ipdb> t_.shape
torch.Size([1, 6, 7080])
```

But for tabular, it seems to be missing a dimension

```
ipdb> (t_)
        [0.9484, 0.0516],
        [0.9642, 0.0358],
        [0.9744, 0.0256],
        [0.9719, 0.0281],
        [0.9803, 0.0197],
        [0.9605, 0.0395],
...
        [0.8389, 0.1611],
        [0.6571, 0.3429],
        [0.2968, 0.7032],
        [0.9422, 0.0578],
        [0.9649, 0.0351]])
ipdb> t_.shape
torch.Size([64, 2])
```

I can confirm this is missing by adding a [None] and getting the results we expect to see: 
```
ipdb> nested_reorder(t_[None], idxs)
tensor([[[0.9017, 0.0983],
         [0.6036, 0.3964],
         [0.1391, 0.8609],
         [0.6555, 0.3445],
         [0.6673, 0.3327],
         ...
```

`t` maps back in get_preds to res and following that back, I see cb.all_tensors() creates res.  When I step into that, I see self.preds is what populates that.  

Still looking into this, but at the moment I think it might have to do with how DataLoader.get_idxs is interacting with the the preds.

**Comment by kevinbird15** [comment]
New fix also would close #3579 if it is a valid change

File: fastai/data/load.py [hunk 2]
```diff
@@ -95,7 +95,12 @@ def __len__(self):
 
     def get_idxs(self):
         idxs = Inf.count if self.indexed else Inf.nones
-        if self.n is not None: idxs = list(itertools.islice(idxs, self.n))
+
+        if self.n is not None:
+            if self.drop_last:
+                idxs = list(itertools.islice(idxs, (self.n//self.bs)*self.bs))
```
[the remark this answers is not in the document]
Comment (kevinbird15) [reply fastai/data/load.py:101]: I think you are correct.  I will write up a test to check this and I think I can move shuffle above the drop last logic

[the remark this answers is not in the document]
Comment (kevinbird15) [reply fastai/data/load.py]: Ok, I believe the new version should address this concern.

**Comment by jph00** [comment]
@kevinbird15 is this ready to merge now, do you think?

**Comment by kevinbird15** [comment]
I would hold off.  I think @marii-moe brought up some good points and I am currently re-considering a lot of the changes I tried making to get_idxs.  I am going to move back a few commits and probably just leave it at making changes to get_preds and leave everything else how it is currently since it has such big implications. I will make the changes back now and let you know when it's done

**Comment by kevinbird15** [comment]
@jph00 This is good to be merged now I believe.  Thanks for the patience hopefully you didn't get too many notifications on this.  I will try coming with a more complete solution next time before I PR. :)

File: fastai/data/load.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kevinbird15) [reply fastai/data/load.py]: Ended up removing the attempt to add drop_last logic to get_idxs and kept everything in get_preds for now.

**Comment by jph00** [comment]
Nice and simple - good one!

**Pull request merged**